### DISCUSSION TITLE
reducing compile time deps

### DISCUSSION THREAD

**Discussion by chungwong** [swoosh/phoenix_swoosh#193] [Ideas]
I have a few files are having their own module for sending out emails(modular structure)

```
lib/hello_world/a/email.ex
lib/hello_world/b/email.ex
lib/hello_world/c/email.ex
lib/hello_world/d/email.ex
...
```
All of those files contain 
```elixir
 use Phoenix.Swoosh, view: HelloWordWeb.EmailView, layout: {HelloWordWeb.LayoutView, :email}
```
Before removing `use Phoenix.Swoosh`, it would take 

```bash
chung@debian:$ time mix compile
Compiling 33 files (.ex)

real	0m8.528s
user	0m19.535s
sys	0m0.990s
```

after removing `use Phoenix.Swoosh`
```bash
chung@debian:$ time mix compile
Compiling 24 files (.ex)

real	0m8.206s
user	0m16.695s
sys	0m0.888s
```
it would reduce the compile time by 1-2 seconds.

I tried to resort to `put_layout/2` and `put_view/2`. However, `use Phoenix.Swoosh` requires a view to be defined, so 
`use Phoenix.Swoosh, view: HelloWordWeb.EmailView` is the minimal requirement. With this change combined with `put_layout`, surprisingly the compile time didn't make a difference at all.

**Comment by princemaple** [comment]
The view is only kept as a name/alias, there is no function call on the view module so there shouldn't be any compile time dependency. Have you done any in-depth profiling & comparison?

**Reply by chungwong** [reply]
I haven't done any profiling(have no clue how to get it started).

However, the only comparison I have done was mentioned above(removing `use Phoenix.Swoosh, view: HelloWordWeb.EmailView`).

The compile time difference in my comparison could be because the compiling time actually fluctuated between 1-2 seconds. That is my pure guess though.

Does it mean although there is a compile time dependency, the overhead will be very very small? just like what `behaviour` is doing?

**Reply by princemaple** [reply]
The impact of compile time dependency is only about re-compilation. A compile time dependency causes extra recompilation when a dependency is changed. During a full compile, it makes no difference.
